doc/dev/decisionrecords/RecordsPOJOsBuilders.md

```diff
@@ -0,0 +1,112 @@
+## Records, POJOs and Builders
+
+We prefer immutable typesafe types over flexibility and "short" class definitions. This makes
+the code more robust and less error-prone. References to other entities might need to be mutable,
+if so try to init them once, and throw an exception if set again. Example:
+
+```java
+Builder initStop(Stop stop) {
+   this.stop = requireNotInitialized(this.stop, stop);
+}
+```
+
+
+### Records
+
+You may use records, but avoid using records if you can not encapsulate it properly. Be especially
+aware of arrays fields (can not be protected) and collections (remember to make a defensive copy).
+If you need to override `equals` and `hashCode`, then it is probably not worth it.
+Be aware that `equals` compare references, not the value of a field. Consider overriding `toString`.
```

Comment:
Are you sure that a record's `equals` use reference equality? https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/Record.html#hashCode()

> If the component is of a reference type, the component is considered equal if and only if [Objects.equals(this.c(), r.c()](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Objects.html#equals(java.lang.Object,java.lang.Object)) would return true.

Reply:
You are right this could be changed, the record uses a shallow eq/hc. Nested fields should also hava a value based eq/hc.

Reply:
I interpret this the same way as @leonardehrenfried: a record's implicit `equals()` method behaves as if it's calling `Obejcts.equals()` on every component of the record, not as if it's comparing them with the `==` operator. When any types are in use that provide semantic equality via an `equals()` override, implicit record equality is in fact deep. For example, a record consisting of nested records, themselves consisting of Strings would automatically have deep semantic equality.

Reply:
@abyrd The doc is changed. Does your comment apply to the old or new version of it?

Reply:
My comment was in response to the version of the text that I saw in the "conversation" tab of the PR, which is now outdated. But the revised text now says: "The default `equals()` and `hashCode()` implementation is shallow, so all nested fields need to be records or value-objects." and I am still not sure about this characterization. 

Consider a record type containing two `List<String>` fields. Comparing two instances of this record type using the record's implicit `equals()` method would behave like calling `Object.equals()` on each field of the record. The default equals method on `List`s applies the `equals` method to each element, and the default `equals` method of `String`s compares their contents rather than identity. So using nothing but default `equals` methods, this record has deep semantic equality. Yet it is not composed of records or value objects, it is composed entirely of standard library collections.

I'm not entirely sure the meaning of the sentence is wrong, it might just be ambiguous or incomplete and need clarification. First I think we need to clarify which "default" the text is referring to - the defaults supplied for all Objects, or the implicit methods generated for record types. Then we need to clarify the intended meaning of "shallow" and "value-objects". The methods in question may technically perform a shallow traversal, but because the methods are recursive and virtual, the method definitions of many standard library classes effectively yield a deep traversal of the entire object graph. Value-objects would seem to exclude reference types, but the widely used `Collections` and `Maps` are reference types, and even if they were nested several layers deep under a record, they would apparently yield deep equality.